org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/DownloadSourcesJob.java

```diff
@@ -339,17 +335,14 @@ private File download(ArtifactKey artifact, List<ArtifactRepository> repositorie
 
   private void scheduleDownload(IProject project, IPackageFragmentRoot fragment, ArtifactKey artifact,
       boolean downloadSources, boolean downloadJavadoc) {
-    addDownloadRequest(project, fragment, artifact, downloadSources, downloadJavadoc);
-
-    schedule(SCHEDULE_INTERVAL);
-  }
-
-  public void addDownloadRequest(IProject project, IPackageFragmentRoot fragment, ArtifactKey artifact,
-      boolean downloadSources, boolean downloadJavadoc) {
     if(project == null || !project.isAccessible()) {
       return;
     }
-    queue.add(new DownloadRequest(project, fragment, artifact, downloadSources, downloadJavadoc));
+    DownloadRequest request = new DownloadRequest(project, fragment, artifact, downloadSources, downloadJavadoc);
+    if(requests.add(request)) { // guard against new requests that are/will be already downloaded in this run to prevent endless loops
```

Comment:
Could we start by something like `!queue.contains(request)` ? Wouldn't that already deliver a noticeable improvement without too much code complexity?

Reply:
Unfortunately that would not help in the case described in https://github.com/eclipse-m2e/m2e-core/issues/252#issuecomment-903173237, because the new request is added when `DownloadSourcesJob.updateClasspath()` is called in `DownloadSourcesJob.run()`. But at that point the queue is already empty, so it would be added again.

That's the reason why the requests are cleared after `updateClasspath()` is called.

Alternatively we could either
- prevent `BuildPathManager.updateClasspath()` from creating new requests (but this would probably require more flags and also adds complexity).

or 
- simply clear the queue after `updateClasspath()` returned in `DownloadSourcesJob.run()`.
But, if any legitimate cases exists (do you know any?) where `updateClasspath()` creates follow-up downloads that downloads would be suppressed.